### DISCUSSION TITLE
Composite modules

### DISCUSSION THREAD

**Discussion by mrpsiho** [builderius/builderius#47] [Ideas]
**Idea**
The idea is to give the possibility to create complex modules or "composite modules".

**What is this**
Composite module can be a layout (several modules) or a single module in terms of structure.

**Benefits**
I see it like saved module/layout with specific settings, which means it is modified in certain way. And then it is accessible where regular modules are, which means easily/fast accessible for using. They are kind of existing saved layouts, but the catch is that composite modules/layouts are not necessary of smth visual. Use case: create a composite module called "Section wrapper" based on Generic block with specific CSS settings like display: flex and maybe width: clamp(...). Later one would use it 7-10 times per template, for each section etc. Time saver!))
_For future - we could connect user roles to specific modules' groups/sets, so that specific user will load builder with a custom set of available modules, including custom ones (so, it might be a mix)._

**What about current saved modules/layouts functionality?**
I do not know what to do with saved modules. Any advice?
As for saved layouts - they should  be transformed into full design assets libraries like other visual tools have. Including connection to central cloud storage etc. The benefit of having such library is to have visual representation of layouts and even templates! You can preview how they look like because of big thumbnails. Saved layouts grouped by type or industry etc. This type of entities is usually used once-twice PER SITE. Like, "ok, I like the design of this template/layouts, let me import it inside my Builderius template and I will proceed working with it then". Whereas composite modules/layouts are meant for much more frequent use.

**Comment by thisbit** [comment]
@mrpsiho I do not really understand the difference between the Layout and compound modules. For instance, i have saved my a11y nav as a layout (is it also not a compound module?)

Also I do not think there is a reason to limit layouts to whole pages or sections. If anything allow to group them ... maybe via tags or something.

And then users can organize them the way it makes sense to them.

No?

**Reply by mrpsiho** [reply]
_Let's imagine we have composite modules functionality._
Having a11y nav menu like yours I would do this: 
1. I would use CSS vars for colors and underline styling with simple black&white fallbacks and saved everything as a composite module. This way I can easily re use it on another project (especially when we add export/import composite modules).
2. I would create a nice looking nav menu based on the design for the client, so I would add specific colors and other design related CSS... let's say I would create cool purple a11y nav menu. Then I would save it as saved layout, with tags: "purple", "a11y nav menu".

See? No limits, I can save the same structure as saved layout and as composite module. It just serves different purpose.
-> Now next time when I have a client with the same nav menu design and purple color - I pick my saved layout and it is done! 
-> Next time when I have a client with different design for nav menu - I pick my composite module, add project specific colors and adjust the menu structure and it is done!

**Comment by mrpsiho** [comment]
@thisbit There is no contradiction between these two functionalities and we do not plan to create any contradiction as well.
Yes, composite modules and saved layouts are **the same in terms of structure**. 

The only difference is that saved layouts are meant to be pieces of design sets. And acts like starters. This is their purpose. Like in your video about nav menu - you have imported the default header and then changed its design based on your needs. Saved layouts are meant to be completely designed, finished pieces. Designer creates beautiful design for templates/layouts, then developer import them and most likely re-design according to the needs of the specific client.
1. _Common usage pattern: you have saved layout with styling -> you import it and remove some DESIGN styles, and new DESIGN and FUNCTIONAL styles_
2. _Saved layouts are inside the Library modal window._
3. _Saved layouts are focused on design, they have preview thumbnails_

Now, composite elements are focused on functionality, let's say. Regular modules are like bricks, but composite modules are special bricks. Tailored bricks, for special purpose. I think 80% of cases composite modules will not have any design styles added, only functional or only structure. Example from bricksbuilder: now they have section, container, block and div modules, right? So, their div is the same as our generic block. Their container is div with width: 1100px. People use containers? Probably yes. But Builderius does not have it at the beginning. Yes, it is very easy to create one: pick generic block and add width: 1100px;display: flex; justify-content:center. BUT! You would have to do this again, and again, and again when you use such container type of div. So, composite module could save time for you: you have to create "Container" composite module once and then just pick it up like you do with generic block. Again, instead of every time adding at least three CSS props (how many click is this? 16 in total?) you would have to make 1-2 clicks and your container is added!
1. _Common usage pattern: you have composite module -> you import it and add new FUNCTIONAL and DESIGN styles_
2. _Composite modules should be where regular modules are_
3. _Composite modules are focused on functionality, they do not have preview thumbnails_

**Comment by thisbit** [comment]
Hmm. Maybe you are right. In generateblocks  I have three ways of using "layouts" which they call Local Patterns.

- i have blank building blocks, grids with proper gaps, column widths etc. Base elements I need to build something.
- finished sections and entire pages even. 
- smaller finished "widgets" which add a small functionality, for example an accordion

The advantage? Everything is in one place, and not mixed with blocks for instance (modules in builderius).

Disadvantages? If I let the client use layouts to build a page with predefined sections, i cannot hide the blank building blocks from them. Because it is all bundled together. 

For instance in my second video. I have used a section with a centered div as layout, in the way you describe modules should be used. Would it be better that these are modules? Maybe. But in the modal you have the modules tab, why not there instead with the default ones?

Now i remember,  Bricks has this functionality, when you create the first one, it has a section on the bottom of the sidebar, called custom modules, so they are separate from the main ones.

You may be right on this one.
